### DISCUSSION TITLE
Is there anyway to manually set a view controller as the root presentation or setting the MasterNavigationController?

### DISCUSSION THREAD

**Discussion by winstonpang** [MvvmCross/MvvmCross#4093] [Q&A]
So I'm dealing wtih an iOS Extension.

I have somehow found a way to start MVVMCross up however the issue is that iOS Share Extensions starts up a Storyboard which then calls your ViewController's logic. Which is all fine, but if you want to make use of the IMvxNavigationService, this serves as a problem, as you now have this view controller that was instantiated by the extension startup process, as a result, trying to reuse some of our external class libraries that heavily rely on the IMvxNavigationService is not working and complains about 

'PresentationAttribute nor MasterNavigationController found for . Assuming Root presentation'

**Comment by Cheesebaron** [comment]
You can always customize the Presenter to do whatever you want. What have you tried?

**Reply by winstonpang** [reply]
I'm not entirely familiar with how to customizer the Presenter.

So my current setup is, iOS Extension Project startsup with a Storyboard, and the only "Entry Point" code is the ViewController (called ShareViewController in our project).

On ViewDidLoad, I am initializing MVVMCross like so:

```            MvxIosSetupSingleton.EnsureSingletonAvailable(new StubAppDelegate(), View?.Window ?? new UIWindow(UIScreen.MainScreen.Bounds)).EnsureInitialized();

            if (Mvx.IoCProvider.TryResolve(out IMvxAppStart startup) && !startup.IsStarted)
            {
                startup.Start(null);
            }
```

Hits all the setup points and it also hits the NavigateToFirstViewModel, however NavigateToFirstViewModel, encounters the error of 'PresentationAttribute nor MasterNavigationController found for . Assuming Root presentation'.

Which is I believe it's because we already have a UIViewController shown from the Storyboard that starts up. 

So I tried something else and on ShareViewController's ViewDidAppear was:

```
var provider = Mvx.IoCProvider.Resolve<IProvideMainListView>();
var viewController = provider.InitializeView() as UIViewController;

NavigationController.PushViewController(new UINavigationController(viewController), true);

```

The IProvideMainListView's InitializeView is simply doing this:

```
            var vmRequest = MvxViewModelRequest.GetDefaultRequest(typeof(MainListViewModel));
            var viewController = new MvxViewController();
            var mainListViewController = viewController.CreateViewControllerFor<MainListViewModel>(vmRequest) as MainListView;

            mainListViewController .OnViewCreate();

            return mainListViewController ;

```


So that does display the List correctly, but when we tap on an item to navigate to another sub page it still encounters the message of

 'PresentationAttribute nor MasterNavigationController found for . Assuming Root presentation'.

I don't know exactly what NavigateToFirstViewModel is doing special, or how it assigns the MasterNavigationController as the designated controller for the IMvxNavigationService.

**Reply by Cheesebaron** [reply]
Override `CreateViewPresenter` in `Setup.cs` and provide your own logic.

Anyway, I am not too familiar with extensions. However, the iOS presenter expects there to be a `RootViewController` assigned to the `Window.RootViewController` that acts as a NavigationController. You need to ensure something like this for your extension.

**Reply by winstonpang** [reply]
Thanks, I got sort of close, but it's just strange.

Our main iOS App that uses MVVMCross is structured like so:

NavigateToFirstViewModel opens up a Main controller that's a MvxTabViewController with the MvxRootPresentation decorated on it.

One of the tabs is a MvxTableViewController (Lets call it my table list), which has no Presentation attributes on it, and when you tap on an item in the table it goes to another view that is a MvxViewController (Lets call it my child view), it also has no Presentation attributes on it.

TableList and navigated to ChildView without any problems at all in the Main iOS App.

In our ShareViewController, which is basically a simple UITableViewController wrapped in a UINavigationController in this extension, on the ViewDidLoad I call this:

```
        private void InitializeMvx()
        {
            var window = View?.Window ?? new UIWindow(UIScreen.MainScreen.Bounds);
            window.RootViewController = NavigationController;
            var appDelegate = new StubAppDelegate();
            appDelegate.Window = window;
            MvxIosSetupSingleton.EnsureSingletonAvailable(appDelegate, window).EnsureInitialized();

            if (Mvx.IoCProvider.TryResolve(out IMvxAppStart startup) && !startup.IsStarted)
            {
                startup.Start(null);
            }
        }

```

Then on ShareViewController's ViewDidAppear.

Mvx.IoCProvider.Resolve<IMvxNavigationService>().Navigate<TableListViewController>();

This shows the TableListViewController that's also used in the Main iOS App.

But for some reason when I tap on the item in the list, it fails to navigation citing No MasterNavigationController, assuming Root Presentation.

So what I ended up adding was MvxChildPresenter on the ChildView, and that seemed to fix it, but it feels a bit voodoo to me.

I have also gone ahead and subclass the MvxIosViewPresenter:

```
        protected override IMvxIosViewPresenter CreateViewPresenter()
        {
            return new MvxIosExtensionViewPresenter(this.ApplicationDelegate, this.Window);
        }
```
But don't have much of an idea which method inside the MvxIosViewPresenter I can override to control the behaviour.
